packages/world/src/SystemCall.sol

```diff
@@ -41,18 +42,26 @@ library SystemCall {
     // Allow access if the system is public or the caller has access to the namespace or name
     if (!publicAccess) AccessControl.requireAccess(resourceSelector, caller);
 
+    // If the msg.value is non-zero, update the namespace's balance
+    if (value > 0) {
+      bytes16 namespace = resourceSelector.getNamespace();
+      uint256 currentBalance = Balances.get(namespace);
+      Balances.set(namespace, currentBalance + value);
+    }
```

Comment:
anything to be worried about here wrt reentry?

Reply:
yes this smells like a re-entry attack

Reply:
Can you elaborate how a reentry attack would work here?

This is a utility that is used by `call`, `callFrom` and the World's fallback function, in each case just passing the `msg.value` to the `SystemCall` util. If the contract that's called in here would call one of these entry points again, it would still only pass that new call's `msg.value` to the function.

I think we need to update the balance table value before calling the system, because the system might want to transfer the balance somewhere else in this call.

Reply:
Yes, `requireAccess` only checks an internal table.

Agree this is definitely a critical part of the system that we should be extra careful with and should also point this out as a focus area in the audit.